support text which already includes tone marks

dound opened the pull request.

oleglegun commented:
Thanks, maybe it makes sense, but it **would be better if it could auto convert tone marks in words to numbers while parsing**.
I don't want to introduce any API changes like `null` values if tone marks can be eventually parsed as numbers.

E.g. 
```
Nǐ -> ['Ni', 3]
bu4yào -> ['bu', 4], ['yao', 4]
```
not
```
Nǐ -> ['Nǐ', null]
bu4yào -> ['bu', 4], ['yào', null]
```

I was writing this lib for pinyin texts serialization to DB for further processing and the goal was to completely get rid of tone marks.

dound commented:
@oleglegun Incorporated your suggestion to auto convert tone marks in words to numbers while parsing.

oleglegun commented:
Ok, based on your proposed changes - it works that way:

```js
// From tests
{
  options: {},
  text: 'Yě kěyǐ.',
  result: [['Yě', 3], ' ', ['kěyǐ', 3], '.']
//            ^               ^ ^
}
```

I personally think it is supposed to work this way:

```js
{
  options: {},
  text: 'Yě kěyǐ.',
  result: [['Ye', 3], ' ', ['ke', 3], ['yi', 3], '.']
//            ^               ^          ^
}
```

Result  array must be clean and do not contain any tone marks. If we need them, we can inversely process the results using [pinyin-tone-convert](https://github.com/oleglegun/pinyin-tone-convert) and get them back. 

The whole process of converting tone marks must be logical. If we convert from marks - we have only numbers, otherwise we have only marks, but not numbers.

So if you implement this functionality I will gladly merge and publish it. Or you can just freely use your own forked repo.

dound commented:
1. Removing the tone marks in the output is straightforward.
1. Unfortunately, splitting up syllables is imo beyond the scope of this code (e.g., figuring out how to split up kěyǐ). There are entire libraries (e.g., [pinyin-split](https://www.npmjs.com/package/pinyin-split)) dedicated to this.

What do you think about this PR only doing (1)?

dound commented:
That done, these changes don't play nicely with pinyin-tone-convert. With #2 (that does not label syllables which already have tone marks and does not strip those tone marks), pinyin-tone-convert can correctly process strings which include some tones already marked.

dound committed:
support text which already includes tone marks

dound committed:
add option to remark pinyin which already has tone marks

remarking removes the tone mark and labels the pinyin with the tone that
was removed (instead of the default which leaves the tone mark in place
and does not label the pinyin)

dound commented:
I've added a new option `remark` which achieves the behavior you requested (stripping tone markings and labeling those syllables with a tone number). (No splitting though.) The default behavior will be to leave the tone marks intact and label the syllables as `null`-- this allows pinyin-tone-convert ([updated here](https://github.com/dound/pinyin-tone-convert/commit/5d39a1eeeab36ab1da88500b048d7976ffbbb112)) to correctly work on strings which already include some (or all) tone markings.

oleglegun commented:
You know what, I think the way to solve this problem is:

1. Add an option `allowToneMarks` (default: false). 
- So, by default, if the library detects any token (like `kěyǐ`) with diacritic marks, it throws an `ToneMarkDetectedError`.
- If we set `{ allowToneMarks: true }` - the library will just set the parsed tone number of `null` leaving all tone marks in place - as you've already implemented: `['kěyǐ', null]`.

2. Add an option `parseToneMarks` (default: false).
- So in case we set `{ allowToneMarks: true, parseToneMarks: true }`, the library will use an external dependency like you mentioned `pinyin-split` to split the token `'kěyǐ' -> ['kě', 'yǐ']` and then we just process it as any valid input to get the final result `[['ke', 3], ['yi', 3]]`.

It think that way of parsing pinyin will be the most flexible and logical one. What do you think?
